### PR TITLE
[Rejection sampling] Add reference score; mason utility

### PR DESCRIPTION
**Pull request by vwxyzjn**
This PR adds the computation for reference scores.

Here is an example dataset: https://huggingface.co/datasets/vwxyzjn/rejection_sampling_31313

This PR also improves workflows with mason, grouping all the shards into the same beaker job, so viewing all the shards and dealing with failures is a bit easier. E.g., [this job](https://beaker.org/ex/01J4X1MNSYVK5KYATSM8FDS46T/tasks/01J4X1MNT7GA81PAX35J9SERBA/job/01J4X1MQ8MS0R9VPQ0X6M2B73F)

image

### CODE REVIEW

File: open_instruct/rejection_sampling.py [hunk 1]
```diff
@@ -65,47 +68,79 @@ def process_shard(rank: int, model_name_or_path: str, args: Args, shard: List[st
...
     tokenizer.add_special_tokens({"pad_token": "[PAD]"})
 
     # Convert the list of data items (shard) into a Hugging Face Dataset object
-    ds = Dataset.from_list(shard)
+    raw_ds = Dataset.from_list(shard)
     # Apply a tokenization function to each item in the dataset
-    ds = ds.map(lambda x: {"input_ids": tokenizer.apply_chat_template(x["messages"])}, remove_columns=ds.column_names)
-
+    ds = raw_ds.map(
+        lambda x: {"input_ids": tokenizer.apply_chat_template(x["messages"])}, remove_columns=raw_ds.column_names
```
Comment (nouhadziri) [inline open_instruct/rejection_sampling.py:81]: Aren't we supposed to score `model_completion` here?
```
reference_completion_ds = raw_ds.map(
        lambda x: {
            "input_ids": tokenizer.apply_chat_template(
                x["messages"][:-1] + [{"content": x["model_completion"], "role": "assistant"}]
            )
        },
        remove_columns=raw_ds.column_names,
    )

File: open_instruct/rejection_sampling.py [hunk 2]
```diff
@@ -65,47 +68,79 @@ def process_shard(rank: int, model_name_or_path: str, args: Args, shard: List[st
...
         torch_dtype=torch.bfloat16,
         attn_implementation="flash_attention_2",
     )
     model = model.to(device)
+    model.eval()
 
     # Initialize a data collator to handle dynamic padding of input sequences
     data_collator = DataCollatorWithPadding(tokenizer=tokenizer)
-    current_batch_size = args.max_forward_batch_size
+    scores = batch_processing_scores(args.max_forward_batch_size, device, tokenizer, ds, model, data_collator)
```
Comment (nouhadziri) [inline open_instruct/rejection_sampling.py:105]: Good refactoring!

**Review by nouhadziri — APPROVED** [review]
Hi Costa, thanks for refactoring, adding reference scores and improving workflows with mason:) I think we're missing the model's completion scores. We're currently only scoring x["messages"].

File: open_instruct/rejection_sampling.py (hunk 1 above)
Comment (vwxyzjn) [reply open_instruct/rejection_sampling.py:81]: https://github.com/allenai/open-instruct/blob/fce3f5667ffe35a75fca9d8d563a163fb00bbc07/open_instruct/generation.py#L115-L117

The `messages` column has the model completion :)

**Comment by vwxyzjn** [comment]
Merging as is.
